aws/signer/v4/middleware.go

```diff
@@ -144,3 +144,35 @@ func SetPayloadHash(ctx context.Context, hash string) context.Context {
 	ctx = context.WithValue(ctx, payloadHashKey{}, hash)
 	return ctx
 }
+
+// AddUnsignedPayloadMiddlewares adds UnsignedPayloadMiddleware to the operation middleware stack
+func AddUnsignedPayloadMiddlewares(stack *middleware.Stack) {
+	unsignedPayloadMiddleware := UnsignedPayloadMiddleware{}
+	stack.Finalize.Add(&unsignedPayloadMiddleware, middleware.After)
+}
+
+// HTTPSignerMiddlewareConfig interface for HTTP signer middleware config
+type HTTPSignerMiddlewareConfig interface {
+	GetSigner() HTTPSigner
+}
+
+// HTTPSignerMiddlewares represent the middleware's for HTTPSigner
+type HTTPSignerMiddlewares struct {
+	computePayloadSHA256Middleware *ComputePayloadSHA256Middleware
+	signHTTPRequestMiddleware      *SignHTTPRequestMiddleware
+}
+
+// AddHTTPSignerMiddlewares adds HTTP signer middleware's to operation stack
+func AddHTTPSignerMiddlewares(stack *middleware.Stack, cfg HTTPSignerMiddlewareConfig, optFns ...func(*HTTPSignerMiddlewares)) {
+	computePayloadSHA256Middleware := ComputePayloadSHA256Middleware{}
+	signHTTPRequestMiddleware := NewSignHTTPRequestMiddleware(cfg.GetSigner())
+	for _, fn := range optFns {
+		fn(&HTTPSignerMiddlewares{
+			computePayloadSHA256Middleware: &computePayloadSHA256Middleware,
+			signHTTPRequestMiddleware:      signHTTPRequestMiddleware,
+		})
+	}
```

Comment:
probably want to create a variable for the `HTTPSignerMiddlewares` value first, then iterate through the `optFns` with a pointer to it. Will prevent creating a new `HTTPSignerMiddlewares` each iteration. i.e initialize `HTTPSignerMiddlewares` once, then iterate with the variable pointer.

Also should only really need to take the address of `computePayloadSHA256Middleware` once. Looks like that is being done multiple times here.  Creating the variable original as an pointer will simplify this.

```go
 	computePayloadSHA256Middleware := &ComputePayloadSHA256Middleware{} 
````

Reply:
updated to
```
// AddHTTPSignerMiddlewares adds HTTP signer middleware's to operation stack
func AddHTTPSignerMiddlewares(stack *middleware.Stack, cfg HTTPSignerMiddlewareConfig, optFns ...func(*HTTPSignerMiddlewares)) {
	m := HTTPSignerMiddlewares {
		ComputePayloadSHA256Middleware: &ComputePayloadSHA256Middleware{},
		SignHTTPRequestMiddleware:      NewSignHTTPRequestMiddleware(cfg.GetSigner()),
	}
	for _, fn := range optFns {
		fn(&m)
	}

	stack.Finalize.Add(m.ComputePayloadSHA256Middleware, middleware.Before)
	stack.Finalize.Add(m.SignHTTPRequestMiddleware, middleware.After)
}



```